### ISSUE TITLE
Form Control: Aria label

### ISSUE THREAD

**Issue by jAigret-Bix** [uswds/uswds#3788]
Display a helpful error message is not available for screen reader users durig navigation to the Drag here fiels. Aria-describedby is not announced by screen readers. It is recommended to use aria-labelledby and associate it with the input form.

![image](https://user-images.githubusercontent.com/67592482/96771252-69ca4400-13af-11eb-8481-8e94691d23bc.png)

**Tasks to complete**
- [X] UX recommendations
- [x] James - first pass on screen reader, verify alerts
- [x] Allie - contact Meg B for next test and monitor results
- [x] Further 508 test on JAWS, verify alerts
- [x] guidance update as needed (not needed at the moment)
- [x] code update as needed

**DOD**
- [ ] all alerts are being read by screen readers
- [ ] guidance and documentation is accurate
- [ ] all tasks complete

**Comment by allie-shaw** [comment]
## Recommendations

- Since aria-labeeledby allows the screen reader to read the label itself, it is recommended to change the aria from aria-describedby to aria-labelledby
- However, since BOTH should be picked up by screen readers, recommend also making sure the code calling for the screen reader works as well.

## Question for Dan
Should we find out why screen readers are not picking detecting “aria-describedby” before changing them all to “aria-labelledby?” Screen readers should be able to detect regardless if it’s describedby or labelledby.

___________________________________________
### ARIA Labels and Relationships

https://developers.google.com/web/fundamentals/accessibility/semantics-aria/aria-labels-and-relationships

aria-labelledby
aria-labelledby allows us to specify the ID of another element in the DOM as an element's label.

This is much like using a label element, with some key differences.
1.	aria-labelledby may be used on any element, not just labelable elements.
2.	While a label element refers to the thing it labels, the relationship is reversed in the the case of aria-labelledby — the thing being labeled refers to the thing that labels it.
3.	Only one label element may be associated with a labelable element, but aria-labelledby can take a list of IDREFs to compose a label from multiple elements. The label will be concatenated in the order that the IDREFs are given.
4.	You can use aria-labelledby to refer to elements that are hidden and would otherwise not be in the accessibility tree. For example, you could add a hidden span next to an element you want to label, and refer to that with aria-labelledby.
5.	However, since ARIA only affects the accessibility tree, aria-labelledby does not give you the familiar label-clicking behavior you get from using a label element.
Importantly, aria-labelledby overrides all other name sources for an element. So, for example, if an element has both an aria-labelledby and an aria-label, or an aria-labelledby and a native HTML label, the aria-labelledby label always takes precedence.
### California Community Colleges
https://cccaccessibility.org/aria-labelledby-vs-aria-describedby-vs-aria-label
- Both can provide an assistive name or description to a user interface control. 
- aria-labelledby -- this is where one element is used to provide a label or an accessible name for another element, the widget element. For example, the aria-labelledby attribute can be used to label a modal so that it's announced to assistive technologies as soon as it's fired.
- aria-describedby-- this is where one element is used to provide an accessible description for another element-- the widget element.
### Medium
https://medium.com/accessibility-a11y/best-practices-on-aria-label-aria-labelledby-and-aria-describedby-913fc9e396ae
- aria-labelledby This attribute allows the label to be set based on content of other ids. 
- aria-describedby This is very similar to aria-labelledby described above. This attribute allows the description to be set for an element based on ids.
- This is hard to tell as both these are very similar. A label describes the essence of an element where as the description provides greater information to help the user. So aria-described by tends to be longer although this is only a heuristic, not a rule.
- Either aria-label or one of aria-labelledby and aria-describedby allow the screen reader to better narrate the inputs of a form to the user and should be used in form input elements. Hence these must be leveraged while developing web forms.

﻿

**Comment by jAigret-Bix** [comment]
Change to aria-labelledby

**Comment by headquarters** [comment]
Hi folks, I wanted to chime in on this because I just ran into a similar situation using some USWDS-based components. Using the Text Input and its corresponding Error Message component, I found that ANDI was picking up the aria-describedby text in the error message but VoiceOver on macOS was not announcing the error message in Safari. 

![Screen Shot 2020-11-04 at 2 00 22 PM](https://user-images.githubusercontent.com/347079/98156703-2808c580-1ea6-11eb-874e-e661e2f2e051.png)

I put together a small test case of just a normal HTML form and tested that to make sure Safari was working as expected:
```
  <div>
    <form action="" method="get">
      <label for="username">Username</label>
      <span id="username-help">Provide a username for us to contact you.</span>
      <input type="text" name="username" id="username" aria-describedby="username-help" />
    </form>
  </div>
```
In this case, the full label and describedby text is announced. 

Next, I tried to remove `role="alert"` from the error message `<span />` and, lo and behold, that worked! Even in Safari's devtools, if you remove the role="alert" attribute, turn on VoiceOver, and focus on the field with an error message everything gets announced properly. 

Here are some guesses for what is happening:
- ARIA Live Regions (i.e. an alert) are expected to exist on the page when it loads and then get populated by text later, so in this case it has content in it already and so it doesn't get announced
- Because this is registered as an alert region, it loses its normal semantics around providing innerText to the AT for announcing the text (wild guess)

I haven't tested this with `aria-labelledby` but will go try that out since I see you all are considering switching things to use that. I would caution against switching attributes to circumvent this bug since it can be fixed other ways. Ideally, `labelledby` would be for short labels and `describedby` would be for lengthier messages, not necessarily one or the other all the time.

**Comment by headquarters** [comment]
Update regarding `aria-labelledby`: if the error message HTML element contains `role="alert"` then even aria-labelledby will fail to announce in VoiceOver. 

```
  <div>
    <form action="" method="get">
      <label for="username">Username</label>
      <span id="username-help" role="alert">Provide a username for us to contact you.</span>
      <input type="text" name="username" id="username" aria-labelledby="username-help" />
    </form>
  </div>
```

Using the same sample form I used before, but with aria-labelledby, I verified that VoiceOver only announced "Username" when the `#username-help` span also included `role="alert"`.

**Comment by mejiaj** [comment]
Voiceover _is_ ignoring both `aria-labelledby` and `aria-describedby`. I was able to get it to read the alert message by removing `role` attribute from alert.

Created a test available [here →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/jm-form-alerts-voiceover/components/preview/input-alerts.html)

### Outstanding questions
- [ ] Are they (left column) being ignored on all screenreaders
- [ ] Test proposed solution (right column)
- [ ] Recommendations or guidance if alerts don't have role="alert". Are we fixing it in Voiceover but breaking everywhere else? What are best practices?

**Comment by allie-shaw** [comment]
Error Messaging (Guidelines and Placement)
https://docs.google.com/document/d/1MiaD96Y1CyvSNKQW3j9wHs1Vig4ZMkN3G2eWNSEpRY8/edit?usp=sharing
